<a href="https://colab.research.google.com/github/AnaMunozz/Analitica-de-Negocios/blob/main/Caso1_Detecci%C3%B3n_de_Riego_SARLAF.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**Caso de Estudio 01**

Desarrollar y analizar un modelo Naive Bayes para la detección de riesgo de lavado de activos (SARLAF), utilizando las información de su comportamiento financiero y los cambios en su patrimonio.
En este sentido, las variables a tener en cuenta son:


* **ingresos_usd:** Ingresos percibidos por el cliente como resultado de sus actividades económicas.

* **gastos_usd:** Total de gastos registrados por el cliente en un período determinado.

* **monto_transado_tarjetas_usd:** Valor total de las transacciones realizadas mediante tarjetas de crédito.

* **porcentaje_crecimiento_patrimonio:** Variación porcentual del patrimonio del cliente, utilizada como indicador de cambios inusuales en su situación financiera.




###0. Carga de librerías
Se procede con la carga de librerías de trabajo

In [ ]:
import numpy as np
import pandas as pd

#Librerías específicas
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import confusion_matrix

###1. Carga de datos
Se cargan los datos de trabajo de la base *2. LavadoActivos*

In [ ]:
nxl = '/content/2. LavadoActivos.xlsx'
XDB = pd.read_excel(nxl)

#Seleccionamos variables de trabajo
XD = XDB [['ingresos_usd', 'gastos_usd', 'monto_transado_tarjetas_usd', 'porcentaje_crecimiento_patrimonio']] # Variables de trabajo
XD.head()


yd = XDB[['lavado_activos']] # Variable de decisión
yd.head()


###2. Implementación Modelo Naive Bayes


In [ ]:
mnb = GaussianNB()
mnb.fit(XD, yd) # Ajustar variables Entrada-Salida

#Mostar las medias de las variables
u = mnb.theta_
sigma = mnb.var_
print('ingresos_usd', 'gastos_usd', 'monto_transado_tarjetas_usd', 'porcentaje_crecimiento_patrimonio')
print(u)
print("\nLas desviaciones son: ")
print(sigma)

ingresos_usd gastos_usd monto_transado_tarjetas_usd porcentaje_crecimiento_patrimonio
[[6.69073255e+03 3.60174970e+03 3.09046404e+03 4.46871435e+01]
 [7.02623043e+04 3.87072844e+04 3.38356197e+04 4.48208630e+01]]

Las desviaciones son: 
[[3.85264064e+07 1.03620619e+07 6.03615435e+06 2.00878669e+02]
 [1.28623418e+10 4.34314704e+09 3.96791535e+09 2.39587239e+02]]


/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:1408: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


# **Análisis de Resultados**
De los resultados arrojados por el modelo, podemos observar que los individuos con riesgo de lavado de activos (*RiesgoLavado*) presentan, en promedio, **ingresos** considerablemente superiores a los del grupo sin riesgo (*NoRiesgo*), lo que muestra la introducción de montos elevados al sistema financiero por pate de estos clientes.

Así mismo, los **gastos** también son mucho mayores en el grupo *RiesgoLavado*, lo que sugiere un nivel de consumo elevado acorde con los mayores ingresos declarados. Además, el el *monto transado a través de tarjetas de crédito* es mayor en los individuos clasificados como *RiesgoLavado*, lo que muestra su alto uso de productos financieros.

Por otro lado, *ambas categorías* presentan **tasas de creimiento del patrimonio** similares en promedio, aunque los *RiesgoLavado* presentan más desviación estandar, siendo un indicador de que, aunque el crecimiento promedio pueda parecer similar, existen casos extremos de incrementos en el patrimonio en el caso de los clasificados como *RiesgoLavado*.


### 3. Evaluación del modelo | Matriz de Confusión
Se procede con la evaluación del modelo. Para este tipo de modelos, se utiliza la matriz de confusión en su evaluación

In [ ]:
ydp = mnb.predict(XD)
cm = confusion_matrix(yd, ydp)
print(cm)

#Métricas de la matriz de confusión
VN=cm[0,0]; FP=cm[0,1]; FN=cm[1,0]; VP=cm[1,1]; TDatos=len(XDB)

#1. Exactitud: Funcionamiento general del modelo
Ex = (VP+VN)/TDatos
print("Exactitud: ", Ex)

#2. Tasa Error: %Fallos del modelo
TEr = (FP+FN)/TDatos
print("Tasa Error: ", TEr)

#3. Sensibilidad: Comportamiento frente a los positivos
Se = VP/(VP+FN)
print("Sensibilidad: ", Se)

#4. Especificidad: Comportamiento frente a los negativos
Es = VN/(VN+FP)
print("Especificidad: ", Es) ##Se preniegan más de los que se preaprueban

#5. Precisión: Versión alternativa del comportamiento frente a los positivos
Pr = VP/(VP+FP)
print("Precisión: ", Pr) ##Con esto, el modelo ya funciona un 80% aprobando créditos

#6. Predicción Negativa: Comportamiento frente a los créditos PreNegados
PrN = VN/(VN+FN)
print("Predicción Negativa: ", PrN)

[[2179   44]
 [ 142  785]]
Exactitud:  0.940952380952381
Tasa Error:  0.05904761904761905
Sensibilidad:  0.8468176914778857
Especificidad:  0.9802069275753487
Precisión:  0.9469240048250904
Predicción Negativa:  0.9388194743644981


##**Análisis de Resultados**

Se evidencia que el modelo presenta un muy buen desempeño en detecctar los riesgo de lavado de activos, alcanzando una **exactitud** de más del 90%, lo cual indica una gran capacidad del modelo para clasificar correctamente a los individuos entre las categorías *RiesgoLavado* y *NoRiesgo*. La **tasa de error** del 5,9% confirma que no se dan muchas clasificaciones incorrectas

Por otro lado, el modelo presenta una **especificidad** del 98%, lo que demuestra su buen desempeño al clasificar correctamente a los individuos sin riesgo de lavado de activos. Este resultado evidencia la capacidad del modelo para minimizar los falsos positivos, asegurando que la mayoría de los clientes legítimos no sean catalogados erróneamente como sospechosos.

En cuanto a la **sensibilidad** (84,7%), se observa que el modelo logra identificar correctamente una muchos de los individuos que presentan riesgo de lavado de activos, pero al ser menos de 90%, muestra de algunos falsos negativos.


Así, estos resultados muestran que el modelo es confiable como herramienta de apoyo para la detección de riesgo SARLAFT, destacándose por su alta *precisión* y *especificidad*. Sin embargo, se recomienda que se utilice con monitoreo continuo.